cpp/oneapi/dal/util/detail/load_graph.hpp

```diff
@@ -18,84 +18,176 @@
+    }
+    delete degrees_vec;
+
+    layout_unfilt->_vertex_neighbors =
+        std::move(vector_vertex_t(rows_cv[layout_unfilt->_vertex_count].get()));
+    layout_unfilt->_edge_offsets = std::move(vector_edge_t(layout_unfilt->_vertex_count + 1));
+    auto _rows_un                = layout_unfilt->_edge_offsets.data();
+    auto _cols_un                = layout_unfilt->_vertex_neighbors.data();
+
+    threader_for(layout_unfilt->_vertex_count + 1, layout_unfilt->_vertex_count + 1, [&](int n) {
+        _rows_un[n] = rows_cv[n].get();
+    });
+
+    threader_for(edges.size(), edges.size(), [&](int u) {
+        _cols_un[rows_cv[edges[u].first].inc() - 1]  = edges[u].second;
+        _cols_un[rows_cv[edges[u].second].inc() - 1] = edges[u].first;
+    });
+    delete rows_vec;
+
+    //removing self-loops,  multiple edges from graph, and make neighbors in CSR sorted
+
+    layout->_vertex_count = layout_unfilt->_vertex_count;
+
+    layout->_degrees = std::move(vector_vertex_t(layout->_vertex_count));
+
+    threader_for(layout_unfilt->_vertex_count, layout_unfilt->_vertex_count, [&](int u) {
+        std::sort(layout_unfilt->_vertex_neighbors.begin() + layout_unfilt->_edge_offsets[u],
+                  layout_unfilt->_vertex_neighbors.begin() + layout_unfilt->_edge_offsets[u + 1]);
+        auto neighs_u_new_end = std::unique(
+            layout_unfilt->_vertex_neighbors.begin() + layout_unfilt->_edge_offsets[u],
+            layout_unfilt->_vertex_neighbors.begin() + layout_unfilt->_edge_offsets[u + 1]);
+        neighs_u_new_end =
+            std::remove(layout_unfilt->_vertex_neighbors.begin() + layout_unfilt->_edge_offsets[u],
+                        neighs_u_new_end,
+                        u);
+        layout->_degrees[u] = neighs_u_new_end - (layout_unfilt->_vertex_neighbors.begin() +
+                                                  layout_unfilt->_edge_offsets[u]);
+    });
+
+    layout->_edge_offsets.clear();
+    layout->_edge_offsets.reserve(layout->_vertex_count + 1);
+
+    total_sum_degrees = 0;
+    layout->_edge_offsets.push_back(total_sum_degrees);
+
+    for (size_t i = 0; i < layout->_vertex_count; ++i) {
+        total_sum_degrees += layout->_degrees[i];
+        layout->_edge_offsets.push_back(total_sum_degrees);
     }
+    layout->_edge_count = layout->_edge_offsets[layout->_vertex_count] / 2;
 
-    free(degrees);
-    layout->_edges.resize(_rows[layout->_vertex_count] + 1);
-    auto _cols = layout->_edges.data();
-    auto offests(layout->_vertexes);
+    layout->_vertex_neighbors =
+        std::move(vector_vertex_t(layout->_edge_offsets[layout->_vertex_count]));
 
-    for (auto edge : edges) {
-        _cols[offests[edge.first]++]  = edge.second;
-        _cols[offests[edge.second]++] = edge.first;
-    }
+    threader_for(layout->_vertex_count, layout->_vertex_count, [&](int u) {
+        auto neighs = oneapi::dal::preview::detail::get_vertex_neighbors_impl(g_unfiltred, u);
```

Comment:
Can we access the vertex neighbors and degrees directly from the layout? Probably we'll save additional calls